<div class="article_content clearfix" id="article_content">
<link href="style.css" rel="stylesheet"/>
<div class="markdown_views prism-atom-one-dark" id="content_views">
<svg style="display: none;" xmlns="http://www.w3.org/2000/svg">
<path d="M5,0 0,2.5 5,5z" id="raphael-marker-block" stroke-linecap="round" style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"></path>
</svg>
<h2><a id="public_0"></a>一、事务方法访问修饰符非public，导致事务失效</h2>
<p><strong>1、实例</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/fbbe0a2b288d495d872742a9ac885eb3.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_14,color_FFFFFF,t_70,g_se,x_16"/><br/> <strong>2、解决</strong><br/> 方式一：将方法修饰符改为public<br/> 方式二：开启AspectJ代理模式</p>
<p><strong>3、注意</strong><br/> 　　如果事务是static、final的，同样无法通过动态代理，事务也是不会生效的。<br/> 　　Spring的声明式事务是基于动态代理实现的，我们无法重写final修饰的方法；<br/> 　　不管是JDK动态代理还是Cglib的动态代理，就是要通过代理的方式获取到代理的具体对象，而static方法修饰的方法是属于类的，不属于任何对象，所以static方法不能被重写，即便写法上是重写，但是并不具备重写的含义，也就是说static方法也不被进行动态代理。</p>
<h2><a id="Transactionalspring_11"></a>二、@Transactional注解的方法抛出的异常不是spring的事务支持的异常，导致事务失效</h2>
<p><strong>1、实例</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/23335bfe41e047a48affd846abde5e9c.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> <strong>2、解决</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/1e9f73c5772842c7b2d69e9be0ff51ca.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> <strong>3、注意</strong><br/> 　　spring的事务只支持未检查异常（unchecked），不支持已检查异常（checked）。</p>
<h2><a id="_18"></a>三、数据表本身是不支持事务，导致事务失效</h2>
<p><strong>1、实例</strong><br/> 　　如果使用MySQL且存储引擎是MyISAM，则事务是不起作用的，原因是MyIASM不支持事务。<br/> <strong>2、解决</strong><br/> 　　数据表可以改为InnoDB存储引擎，支持事务</p>
<h2><a id="Transactionalspring_23"></a>四、@Transactional注解所在的类没有被spring管理，导致事务失效</h2>
<p><strong>1、实例</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/3e0f7b008562459790eef1aedb35fd0f.png"/><br/> <strong>2、解决</strong><br/> 　　加上@Service注解或者使用其他能注册成Spring Bean的方式或注解。</p>
<h2><a id="catch_28"></a>五、catch掉异常之后，没有再次抛出异常，导致事务失效</h2>
<p><strong>1、实例</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/3a74bce5270e48da8ff2048437b476de.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> <strong>2、解决</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/3b3abea786444facabe7cb05d474d901.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> <strong>3、注意</strong><br/> 　　如果在加有事务的方法内，使用了try…catch…语句块对异常进行了捕获，而catch语句块没有throw new RuntimeException异常或者Spring支持的异常类型，则事务不会回滚。</p>
<h2><a id="this_35"></a>六、方法自身（this）调用问题，导致事务失效</h2>
<p>非事务方法insert()中调用的自身类的事务方法insertUser()。<br/> <strong>1、实例</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/419c0187b3054687a146d54e62c8a548.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> <strong>2、解析</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/879d130722cf4731845f0e7f0f4f2fa6.png"/></p>
<p><strong>3、解决</strong><br/> 方式一:<br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/5d6e336ba6ab4de0bfa8655a23f7cb57.png"/><br/> 方式二：<br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/1413c6003ef941a5b7bdcfa8e2f73ffc.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/12e2683c7e0e4c5984393affcd668752.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> 方式三：<br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/7ae2b486fd5c4dee90e64714f0666240.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> 方式四<br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/f31ee1e139184f1db00edbc1b77415c6.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/></p>
<h2><a id="_53"></a>七、数据源没有配置事务管理器，导致事务失效</h2>
<h2><a id="_55"></a>八、传播类型不支持事务，导致事务失效</h2>
<p><strong>1、实例</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/1130bca4d3884510ae71be660def7856.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/></p>
<p><strong>2、注意</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/155deaf2c2d34981b2a43a6f743432d0.png"/></p>
<h2><a id="_62"></a>九、多线程调用，导致事务失效</h2>
<p><strong>1、实例</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/2373bdfd11bd4c1f9f07334d37f25fd6.png?x-oss-process=image/watermark,type_d3F5LXplbmhlaQ,shadow_50,text_Q1NETiBA5ZWK6I27772e,size_20,color_FFFFFF,t_70,g_se,x_16"/><br/> <strong>2、解析</strong><br/> <img alt="在这里插入图片描述" src="https://img-blog.csdnimg.cn/f4e7354a346d4a99a45b561129f83f89.png"/></p>
</div>
<link href="https://csdnimg.cn/release/blogv2/dist/mdeditor/css/editerView/markdown_views-22a2fefd3b.css" rel="stylesheet"/>
<link href="https://csdnimg.cn/release/blogv2/dist/mdeditor/css/style-4f8fbf9108.css" rel="stylesheet"/>
</div>